Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp

```diff
@@ -0,0 +1,238 @@
+ * @param histoTopic The name of the topic streaming the histo data
+ * @param spDetTopic The name of the topic streaming the spectrum-detector
+ * run mapping
+ */
+KafkaHistoStreamDecoder::KafkaHistoStreamDecoder(
+    std::shared_ptr<IKafkaBroker> broker, const std::string &histoTopic,
+    const std::string &instrumentName)
+    : m_broker(broker), m_histoTopic(histoTopic),
+      m_instrumentName(instrumentName), m_histoStream(), m_workspace(),
+      m_buffer(), m_thread(), m_interrupt(false), m_capturing(false),
+      m_exception() {
+  // Initialize buffer workspace
+  m_workspace = createBufferWorkspace();
+}
+
+/**
+ * Destructor.
+ * Stops capturing from the stream
+ */
+KafkaHistoStreamDecoder::~KafkaHistoStreamDecoder() { stopCapture(); }
+
+/**
+ * Start capturing from the stream on a separate thread. This is a non-blocking
+ * call and will return after the thread has started
+ */
+void KafkaHistoStreamDecoder::startCapture(bool) {
+  g_log.debug() << "Starting capture on topic: " << m_histoTopic << "\n";
+  m_histoStream =
+      m_broker->subscribe({m_histoTopic}, SubscribeAtOption::LATEST);
+
+  m_thread = std::thread([this]() { this->captureImpl(); });
+  m_thread.detach();
+}
+
+/**
+ * Stop capturing from the stream. This is a blocking call until the capturing
+ * function has completed
+ */
+void KafkaHistoStreamDecoder::stopCapture() {
+  g_log.debug() << "Stopping capture\n";
+
+  // This will interrupt the "event" loop
+  m_interrupt = true;
+  // Wait until the function has completed. The background thread will exit
+  // automatically
+  while (m_capturing) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  };
+}
+
+/**
+ * Check if there is data available to extract
+ * @return True if data has been accumulated so that extractData()
+ * can be called, false otherwise
+ */
+bool KafkaHistoStreamDecoder::hasData() const {
+  std::lock_guard<std::mutex> lock(m_buffer_mutex);
+  return !m_buffer.empty();
+}
+
+/**
+ * Check for an exception thrown by the background thread and rethrow
+ * it if necessary.
+ * @return A pointer to the data collected since the last call to this
+ * method
+ */
+API::Workspace_sptr KafkaHistoStreamDecoder::extractData() {
+  if (m_exception) {
+    throw *m_exception;
+  }
+
+  auto workspace_ptr = extractDataImpl();
+  return workspace_ptr;
+}
+
+// -----------------------------------------------------------------------------
+// Private members
+// -----------------------------------------------------------------------------
+
+API::Workspace_sptr KafkaHistoStreamDecoder::extractDataImpl() {
+  std::lock_guard<std::mutex> lock(m_buffer_mutex);
+
+  if (!m_capturing) {
+    throw Exception::NotYet("Local buffers not initialized.");
+  }
+
+  if (m_buffer.empty()) {
+    throw Exception::NotYet("No message to process yet.");
+  }
+
+  auto histoMsg = GetEventHistogram(m_buffer.c_str());
+
+  auto shape = histoMsg->current_shape();
```

Comment:
minor point, but do these need to be `const auto` since we only read, no modification required.